plugins/logitech-bulkcontroller/fu-logitech-bulkcontroller-device.c

```diff
@@ -0,0 +1,522 @@
+	init_ack = fu_logitech_bulkcontroller_device_send_upd_cmd(self, CMD_INIT, NULL, error);
+	if (init_ack == NULL) {
+		g_prefix_error(error, "error in writing init transfer packet: ");
+		return FALSE;
+	}
+
+	/* transfer sent */
+	fu_device_set_status(device, FWUPD_STATUS_DEVICE_WRITE);
+	fu_byte_array_append_uint64(start_pkt, g_bytes_get_size(fw), G_LITTLE_ENDIAN);
+	start_ack = fu_logitech_bulkcontroller_device_send_upd_cmd(self,
+								   CMD_START_TRANSFER,
+								   start_pkt,
+								   error);
+	if (start_ack == NULL) {
+		g_prefix_error(error, "error in writing init transfer packet: ");
+		return FALSE;
+	}
+
+	/* each block */
+	chunks = fu_chunk_array_new_from_bytes(fw, 0x0, 0x0, PAYLOAD_SIZE);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autoptr(GByteArray) data_ack = NULL;
+		g_autoptr(GByteArray) data_pkt = g_byte_array_new();
+
+		g_byte_array_append(data_pkt, fu_chunk_get_data(chk), fu_chunk_get_data_sz(chk));
+		data_ack = fu_logitech_bulkcontroller_device_send_upd_cmd(self,
+									  CMD_DATA_TRANSFER,
+									  data_pkt,
+									  error);
+		if (data_ack == NULL) {
+			g_prefix_error(error, "failed to send data packet 0x%x: ", i);
+			return FALSE;
+		}
+		fu_device_set_progress_full(FU_DEVICE(self), i + 1, chunks->len);
+	}
+
+	/* sending end transfer */
+	base64hash = fu_logitech_bulkcontroller_device_compute_hash(fw);
+	fu_byte_array_append_uint32(end_pkt, 1, G_LITTLE_ENDIAN);   /* update */
+	fu_byte_array_append_uint32(end_pkt, 0, G_LITTLE_ENDIAN);   /* force */
+	fu_byte_array_append_uint32(end_pkt, MD5, G_LITTLE_ENDIAN); /* checksum type */
+	g_byte_array_append(end_pkt, (const guint8 *)base64hash, strlen(base64hash));
+	end_ack =
+	    fu_logitech_bulkcontroller_device_send_upd_cmd(self, CMD_END_TRANSFER, end_pkt, error);
+	if (end_ack == NULL) {
+		g_prefix_error(error, "error in writing init transfer packet: ");
+		return FALSE;
+	}
+
+	/* send uninit */
+	uninit_ack = fu_logitech_bulkcontroller_device_send_upd_cmd(self, CMD_UNINIT, NULL, error);
+	if (uninit_ack == NULL) {
+		g_prefix_error(error, "error in writing finish transfer packet: ");
+		return FALSE;
+	}
+
+	/* success! */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_bulkcontroller_device_open(FuDevice *device, GError **error)
+{
+	FuLogitechBulkcontrollerDevice *self = FU_LOGITECH_BULKCONTROLLER_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+
+	/* FuUsbDevice->open */
+	if (!FU_DEVICE_CLASS(fu_logitech_bulkcontroller_device_parent_class)->open(device, error))
+		return FALSE;
+
+	/* claim both interfaces */
+	if (!g_usb_device_claim_interface(usb_device,
+					  self->update_iface,
+					  G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					  error)) {
+		return FALSE;
+	}
+	if (!g_usb_device_claim_interface(usb_device,
+					  self->sync_iface,
+					  G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					  error)) {
+		return FALSE;
```

Comment:
I think it would be good to prefix the errors here in case claiming one of these ends up being problematic you know which one it was